File: src/MessageFlow/HandlerBase.cs
namespace MessageFlow;

/// <summary>
/// Convenience base class for handlers that either fully handle a request or pass it on.
/// </summary>
/// <typeparam name="TRequest">The type of the request flowing through the chain.</typeparam>
/// <typeparam name="TResponse">The type of the response produced by the chain.</typeparam>
public abstract class HandlerBase<TRequest, TResponse> : IHandler<TRequest, TResponse>
{
    /// <summary>
    /// Determines whether this handler is responsible for the given request.
    /// </summary>
    /// <param name="request">The request to inspect.</param>
    /// <returns><see langword="true"/> when this handler should process the request.</returns>
    protected abstract bool CanHandle(TRequest request);

    /// <summary>
    /// Processes a request this handler is responsible for.
    /// </summary>
    /// <param name="request">The request to process.</param>
    /// <param name="cancellationToken">A token used to cancel the operation.</param>
    /// <returns>The response for the request.</returns>
    protected abstract ValueTask<TResponse> ProcessAsync(TRequest request, CancellationToken cancellationToken);

    /// <inheritdoc />
    public ValueTask<TResponse> HandleAsync(
        TRequest request,
        NextHandler<TRequest, TResponse> nextHandler,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(nextHandler);

        return CanHandle(request)
            ? ProcessAsync(request, cancellationToken)
            : nextHandler(request, cancellationToken);
    }
}

File: node/src/HandlerBase.ts
import type { CancellationToken } from './CancellationToken.js';
import type { Handler } from './Handler.js';
import type { NextHandler } from './NextHandler.js';

/**
 * Convenience base class for handlers that either fully handle a request or pass it on.
 *
 * @typeParam TRequest - the type of the request flowing through the chain
 * @typeParam TResponse - the type of the response produced by the chain
 */
export abstract class HandlerBase<TRequest, TResponse>
  implements Handler<TRequest, TResponse>
{
  /**
   * Determines whether this handler is responsible for the given request.
   *
   * @param request - the request to inspect
   * @returns `true` when this handler should process the request
   */
  protected abstract canHandle(request: TRequest): boolean;

  /**
   * Processes a request this handler is responsible for.
   *
   * @param request - the request to process
   * @param cancellationToken - a token used to cancel the operation
   * @returns the response for the request
   */
  protected abstract process(
    request: TRequest,
    cancellationToken: CancellationToken
  ): Promise<TResponse>;

  /**
   * Processes the request, optionally delegating to the next handler of the chain.
   *
   * @param request - the request to process
   * @param next - the next handler of the chain
   * @param cancellationToken - a token used to cancel the operation
   * @returns the response for the request
   */
  handle(
    request: TRequest,
    next: NextHandler<TRequest, TResponse>,
    cancellationToken: CancellationToken
  ): Promise<TResponse> {
    if (next === null || next === undefined) {
      throw new TypeError('next must not be null or undefined');
    }

    return this.canHandle(request)
      ? this.process(request, cancellationToken)
      : next(request, cancellationToken);
  }
}
